File: src/Core.Go/zeta_id/zeta_id.go
package zeta_id

import (
	"errors"
	"fmt"
	"math/big"
)

type AuthorityType string

const (
	AuthorityHumanVerified AuthorityType = "HumanVerified"
	AuthorityTrustedAgent  AuthorityType = "TrustedAgent"
	AuthorityStandard      AuthorityType = "Standard"
	AuthorityBestEffort    AuthorityType = "BestEffort"
	AuthoritySimulated     AuthorityType = "Simulated"
	AuthorityRaw           AuthorityType = "Raw"
)

type Authority struct {
	Type  AuthorityType
	Value uint8
}

type MomentumType string

const (
	MomentumBackground MomentumType = "Background"
	MomentumNormal     MomentumType = "Normal"
	MomentumElevated   MomentumType = "Elevated"
	MomentumHigh       MomentumType = "High"
	MomentumCritical   MomentumType = "Critical"
	MomentumRaw        MomentumType = "Raw"
)

type Momentum struct {
	Type  MomentumType
	Value uint8
}

type ZetaObservation struct {
	Version    uint8
	Timestamp  uint64
	Chromosome uint8
	Category   uint8
	Firefly    uint8
	Authority  Authority
	Persona    uint8
	Momentum   Momentum
	Location   uint8
}

var AuthorityValues = map[AuthorityType]uint8{
	AuthoritySimulated:     3,
	AuthorityBestEffort:    8,
	AuthorityStandard:      15,
	AuthorityTrustedAgent:  20,
	AuthorityHumanVerified: 31,
}

var MomentumValues = map[MomentumType]uint8{
	MomentumBackground: 32,
	MomentumNormal:     96,
	MomentumElevated:   160,
	MomentumHigh:       224,
	MomentumCritical:   248,
}

func isNamedAuthorityByte(val uint8) bool {
	for _, v := range AuthorityValues {
		if v == val {
			return true
		}
	}
	return false
}

func isNamedMomentumByte(val uint8) bool {
	for _, v := range MomentumValues {
		if v == val {
			return true
		}
	}
	return false
}

func AuthorityToByte(a Authority) (uint8, error) {
	if a.Type == AuthorityRaw {
		if a.Value > 31 {
			return 0, fmt.Errorf("Authority.Raw value %d must be 0..31", a.Value)
		}
		if isNamedAuthorityByte(a.Value) {
			return 0, fmt.Errorf("Authority.Raw value %d aliases a named case", a.Value)
		}
		return a.Value, nil
	}
	v, ok := AuthorityValues[a.Type]
	if !ok {
		return 0, fmt.Errorf("unknown authority type %s", a.Type)
	}
	return v, nil
}

func AuthorityFromByte(v uint8) Authority {
	for t, val := range AuthorityValues {
		if val == v {
			return Authority{Type: t, Value: v}
		}
	}
	return Authority{Type: AuthorityRaw, Value: v}
}

func MomentumToByte(m Momentum) (uint8, error) {
	if m.Type == MomentumRaw {
		if isNamedMomentumByte(m.Value) {
			return 0, fmt.Errorf("Momentum.Raw value %d aliases a named case", m.Value)
		}
		return m.Value, nil
	}
	v, ok := MomentumValues[m.Type]
	if !ok {
		return 0, fmt.Errorf("unknown momentum type %s", m.Type)
	}
	return v, nil
}

func MomentumFromByte(v uint8) Momentum {
	for t, val := range MomentumValues {
		if val == v {
			return Momentum{Type: t, Value: v}
		}
	}
	return Momentum{Type: MomentumRaw, Value: v}
}

type SimulationEnvironment interface {
	NextInt64() int64
}

type DeterministicEnv struct{}

func (d DeterministicEnv) NextInt64() int64 {
	return 0
}

type BitField struct {
	Offset uint
	Width  uint
}

var BitLayout = struct {
	Version    BitField
	Timestamp  BitField
	Chromosome BitField
	Category   BitField
	Firefly    BitField
	Authority  BitField
	Persona    BitField
	Momentum   BitField
	Location   BitField
	Randomness BitField
}{
	Version:    BitField{Offset: 123, Width: 5},
	Timestamp:  BitField{Offset: 75, Width: 48},
	Chromosome: BitField{Offset: 70, Width: 5},
	Category:   BitField{Offset: 65, Width: 4},
	Firefly:    BitField{Offset: 64, Width: 1},
	Authority:  BitField{Offset: 59, Width: 5},
	Persona:    BitField{Offset: 51, Width: 8},
	Momentum:   BitField{Offset: 43, Width: 8},
	Location:   BitField{Offset: 35, Width: 8},
	Randomness: BitField{Offset: 0, Width: 32},
}

func setBits(target *big.Int, field BitField, val uint64) {
	valBig := new(big.Int).SetUint64(val)
	mask := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), field.Width), big.NewInt(1))
	valBig.And(valBig, mask)
	valBig.Lsh(valBig, field.Offset)
	target.Or(target, valBig)
}

func getBits(target *big.Int, field BitField) uint64 {
	valBig := new(big.Int).Rsh(target, field.Offset)
	mask := new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), field.Width), big.NewInt(1))
	valBig.And(valBig, mask)
	return valBig.Uint64()
}

func Pack(obs ZetaObservation, env SimulationEnvironment) (*big.Int, error) {
	if obs.Timestamp > ((1 << 48) - 1) {
		return nil, fmt.Errorf("timestamp %d exceeds 48-bit range", obs.Timestamp)
	}
	if obs.Version > 31 {
		return nil, fmt.Errorf("version %d exceeds 5-bit range", obs.Version)
	}
	if obs.Chromosome > 31 {
		return nil, fmt.Errorf("chromosome %d exceeds 5-bit range", obs.Chromosome)
	}
	if obs.Category > 15 {
		return nil, fmt.Errorf("category %d exceeds 4-bit range", obs.Category)
	}
	if obs.Category >= 9 {
		return nil, errors.New("ZetaObservation.Category must be < 9")
	}
	if obs.Firefly > 1 {
		return nil, fmt.Errorf("firefly %d exceeds 1-bit range", obs.Firefly)
	}

	authByte, err := AuthorityToByte(obs.Authority)
	if err != nil {
		return nil, err
	}
	momByte, err := MomentumToByte(obs.Momentum)
	if err != nil {
		return nil, err
	}

	id := new(big.Int)
	setBits(id, BitLayout.Version, uint64(obs.Version))
	setBits(id, BitLayout.Timestamp, obs.Timestamp)
	setBits(id, BitLayout.Chromosome, uint64(obs.Chromosome))
	setBits(id, BitLayout.Category, uint64(obs.Category))
	setBits(id, BitLayout.Firefly, uint64(obs.Firefly))
	setBits(id, BitLayout.Authority, uint64(authByte))
	setBits(id, BitLayout.Persona, uint64(obs.Persona))
	setBits(id, BitLayout.Momentum, uint64(momByte))
	setBits(id, BitLayout.Location, uint64(obs.Location))

	rand32 := uint64(env.NextInt64() & 0xffffffff)
	setBits(id, BitLayout.Randomness, rand32)

	return id, nil
}

func Unpack(id *big.Int) ZetaObservation {
	version := uint8(getBits(id, BitLayout.Version))
	timestamp := getBits(id, BitLayout.Timestamp)
	chromosome := uint8(getBits(id, BitLayout.Chromosome))
	category := uint8(getBits(id, BitLayout.Category))
	firefly := uint8(getBits(id, BitLayout.Firefly))
	authByte := uint8(getBits(id, BitLayout.Authority))
	persona := uint8(getBits(id, BitLayout.Persona))
	momByte := uint8(getBits(id, BitLayout.Momentum))
	location := uint8(getBits(id, BitLayout.Location))

	return ZetaObservation{
		Version:    version,
		Timestamp:  timestamp,
		Chromosome: chromosome,
		Category:   category,
		Firefly:    firefly,
		Authority:  AuthorityFromByte(authByte),
		Persona:    persona,
		Momentum:   MomentumFromByte(momByte),
		Location:   location,
	}
}

func ToHex(id *big.Int) string {
	hexStr := fmt.Sprintf("%x", id)
	for len(hexStr) < 32 {
		hexStr = "0" + hexStr
	}
	return hexStr
}

File: src/Core.Python/src/zeta/zeta_id.py
import secrets
from typing import NamedTuple, Protocol

import zeta.zeta_id_gen as gen

AUTHORITY_VALUES: dict[str, int] = {
    "Simulated": 3,
    "BestEffort": 8,
    "Standard": 15,
    "TrustedAgent": 20,
    "HumanVerified": 31,
}

MOMENTUM_VALUES: dict[str, int] = {
    "Background": 32,
    "Normal": 96,
    "Elevated": 160,
    "High": 224,
    "Critical": 248,
}


class Authority:
    def __init__(self, type_: str, value: int = 0):
        self.type = type_
        self.value = value

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Authority):
            return False
        return self.type == other.type and self.value == other.value

    def __repr__(self) -> str:
        return f"Authority(type={self.type!r}, value={self.value!r})"


class Momentum:
    def __init__(self, type_: str, value: int = 0):
        self.type = type_
        self.value = value

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Momentum):
            return False
        return self.type == other.type and self.value == other.value

    def __repr__(self) -> str:
        return f"Momentum(type={self.type!r}, value={self.value!r})"


class ZetaObservation:
    def __init__(
        self,
        version: int,
        timestamp: int,
        chromosome: int,
        category: int,
        authority: Authority,
        persona: int,
        momentum: Momentum,
        location: int,
    ):
        self.version = version
        self.timestamp = timestamp
        self.chromosome = chromosome
        self.category = category
        self.authority = authority
        self.persona = persona
        self.momentum = momentum
        self.location = location

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ZetaObservation):
            return False
        return (
            self.version == other.version
            and self.timestamp == other.timestamp
            and self.chromosome == other.chromosome
            and self.category == other.category
            and self.authority == other.authority
            and self.persona == other.persona
            and self.momentum == other.momentum
            and self.location == other.location
        )

    def __repr__(self) -> str:
        return (
            f"ZetaObservation(version={self.version!r}, timestamp={self.timestamp!r}, "
            f"chromosome={self.chromosome!r}, category={self.category!r}, "
            f"authority={self.authority!r}, persona={self.persona!r}, momentum={self.momentum!r}, "
            f"location={self.location!r})"
        )


class BitField(NamedTuple):
    offset: int
    width: int


class BitLayout:
    # Every offset below comes straight from the generated constants — this class
    # never advances a running offset, so removing a field cannot shift its
    # neighbours. Named-field widths sum to 123; the other 5 bits are RESERVED.
    # Bit 64 is one of them: it held the 1-bit Firefly field until 2026-08-11,
    # when it was reclaimed NO-SHIFT (Category stays at 65, Chromosome at 70,
    # Timestamp at 75, Version at 123). pack() must leave it zero and unpack()
    # must ignore it; writing it in V1 is an unversioned schema break.
    # See docs/zeta-id-v1-layout.yaml `reserved_bits`.
    version = BitField(gen.VERSION_OFFSET, gen.VERSION_WIDTH)
    timestamp = BitField(gen.TIMESTAMP_OFFSET, gen.TIMESTAMP_WIDTH)
    chromosome = BitField(gen.CHROMOSOME_OFFSET, gen.CHROMOSOME_WIDTH)
    category = BitField(gen.CATEGORY_OFFSET, gen.CATEGORY_WIDTH)
    # (bit 64 — RESERVED, formerly Firefly; deliberately no BitField here)
    authority = BitField(gen.AUTHORITY_OFFSET, gen.AUTHORITY_WIDTH)
    persona = BitField(gen.PERSONA_OFFSET, gen.PERSONA_WIDTH)
    momentum = BitField(gen.MOMENTUM_OFFSET, gen.MOMENTUM_WIDTH)
    location = BitField(gen.LOCATION_OFFSET, gen.LOCATION_WIDTH)
    randomness = BitField(gen.RANDOMNESS_OFFSET, gen.RANDOMNESS_WIDTH)


def set_bits(value: int, field: BitField, field_value: int) -> int:
    mask = (1 << field.width) - 1
    return value | ((field_value & mask) << field.offset)


def get_bits(value: int, field: BitField) -> int:
    mask = (1 << field.width) - 1
    return (value >> field.offset) & mask


class SimulationEnvironment(Protocol):
    def next_int64(self) -> int: ...


class DeterministicEnv:
    def next_int64(self) -> int:
        return 0


class DefaultEnv:
    def next_int64(self) -> int:
        return secrets.randbits(64)


DETERMINISTIC_ENV = DeterministicEnv()
DEFAULT_ENV = DefaultEnv()


def pack(obs: ZetaObservation, env: SimulationEnvironment = DEFAULT_ENV) -> int:
    if obs.timestamp > ((1 << 48) - 1):
        raise ValueError(f"timestamp {obs.timestamp} exceeds 48-bit range")
    if obs.version > 31:
        raise ValueError(f"version {obs.version} exceeds 5-bit range")
    if obs.chromosome > 31:
        raise ValueError(f"chromosome {obs.chromosome} exceeds 5-bit range")
    if obs.category > 15:
        raise ValueError(f"category {obs.category} exceeds 4-bit range")
    if obs.category >= 9:
        raise ValueError("ZetaObservation.category must be < 9")

    # Authority mapping
    if obs.authority.type == "Raw":
        if obs.authority.value > 31:
            raise ValueError(f"Authority.Raw value {obs.authority.value} must be 0..31")
        if obs.authority.value in AUTHORITY_VALUES.values():
            raise ValueError(
                f"Authority.Raw value {obs.authority.value} aliases a named case"
            )
        auth_byte = obs.authority.value
    else:
        if obs.authority.type not in AUTHORITY_VALUES:
            raise ValueError(f"unknown authority type {obs.authority.type}")
        auth_byte = AUTHORITY_VALUES[obs.authority.type]

    # Momentum mapping
    if obs.momentum.type == "Raw":
        if obs.momentum.value in MOMENTUM_VALUES.values():
            raise ValueError(
                f"Momentum.Raw value {obs.momentum.value} aliases a named case"
            )
        mom_byte = obs.momentum.value
    else:
        if obs.momentum.type not in MOMENTUM_VALUES:
            raise ValueError(f"unknown momentum type {obs.momentum.type}")
        mom_byte = MOMENTUM_VALUES[obs.momentum.type]

    bits = 0
    bits = set_bits(bits, BitLayout.version, obs.version)
    bits = set_bits(bits, BitLayout.timestamp, obs.timestamp)
    bits = set_bits(bits, BitLayout.chromosome, obs.chromosome)
    bits = set_bits(bits, BitLayout.category, obs.category)
    # bit 64 is RESERVED (formerly Firefly) — never written, stays zero
    bits = set_bits(bits, BitLayout.authority, auth_byte)
    bits = set_bits(bits, BitLayout.persona, obs.persona)
    bits = set_bits(bits, BitLayout.momentum, mom_byte)
    bits = set_bits(bits, BitLayout.location, obs.location)

    rand = env.next_int64() & 0xFFFFFFFF
    bits = set_bits(bits, BitLayout.randomness, rand)

    return bits


def unpack(id_val: int) -> ZetaObservation:
    version = get_bits(id_val, BitLayout.version)
    timestamp = get_bits(id_val, BitLayout.timestamp)
    chromosome = get_bits(id_val, BitLayout.chromosome)
    category = get_bits(id_val, BitLayout.category)
    # bit 64 is RESERVED (formerly Firefly) — ignored on read
    auth_byte = get_bits(id_val, BitLayout.authority)
    persona = get_bits(id_val, BitLayout.persona)
    mom_byte = get_bits(id_val, BitLayout.momentum)
    location = get_bits(id_val, BitLayout.location)

    auth_type = "Raw"
    for k, v in AUTHORITY_VALUES.items():
        if v == auth_byte:
            auth_type = k
            break
    authority = Authority(type_=auth_type, value=auth_byte)

    mom_type = "Raw"
    for k, v in MOMENTUM_VALUES.items():
        if v == mom_byte:
            mom_type = k
            break
    momentum = Momentum(type_=mom_type, value=mom_byte)

    return ZetaObservation(
        version=version,
        timestamp=timestamp,
        chromosome=chromosome,
        category=category,
        authority=authority,
        persona=persona,
        momentum=momentum,
        location=location,
    )


CROCKFORD_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"
ZETAID_BASE32_LEN = 26
ZETAID_HEX_LEN = 32
MASK_128 = (1 << 128) - 1

DECODE: dict[str, int] = {}
for i, char in enumerate(CROCKFORD_ALPHABET):
    DECODE[char] = i
    DECODE[char.lower()] = i

DECODE["I"] = 1
DECODE["i"] = 1
DECODE["L"] = 1
DECODE["l"] = 1
DECODE["O"] = 0
DECODE["o"] = 0


def format_b32(id_val: int) -> str:
    v = id_val & MASK_128
    out: list[str] = [""] * ZETAID_BASE32_LEN
    for i in range(ZETAID_BASE32_LEN - 1, -1, -1):
        out[i] = CROCKFORD_ALPHABET[v & 31]
        v >>= 5
    return "".join(out)


def parse_b32(s: str) -> int:
    if len(s) != ZETAID_BASE32_LEN:
        raise ValueError(
            f"expected {ZETAID_BASE32_LEN} Crockford-base32 chars, got {len(s)}"
        )
    v = 0
    for i, ch in enumerate(s):
        if ch not in DECODE:
            raise ValueError(f"invalid Crockford-base32 char {ch!r} at index {i}")
        v = (v << 5) | DECODE[ch]
    if v > MASK_128:
        raise ValueError("value exceeds 128 bits")
    return v


def to_hex(id_val: int) -> str:
    h = hex(id_val & MASK_128)[2:]
    return h.zfill(ZETAID_HEX_LEN)


def from_hex(s: str) -> int:
    if len(s) != ZETAID_HEX_LEN:
        raise ValueError(f"expected {ZETAID_HEX_LEN} hex chars, got {len(s)}")
    return int(s, 16) & MASK_128


def is_canonical(s: str) -> bool:
    if len(s) != ZETAID_BASE32_LEN:
        return False
    for ch in s:
        if ch not in CROCKFORD_ALPHABET:
            return False
    first_char = s[0]
    first_val = DECODE.get(first_char)
    if first_val is None or first_val >= 8:
        return False
    try:
        return format_b32(parse_b32(s)) == s
    except ValueError:
        return False
